---
title: "RVC_PCA_analysis"
author: "Stephanie Goya"
date: "2023-12-18"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Upload the dataframes and alignment (in FASTA) to calculate the pairwise genetic distance matrix:
```{r dataframes, echo=TRUE}
library("ape") #pairwise distance from DNA sequences
library("seqinr") #read alignment 
library("FactoMineR") #Principal Component Analysis (PCA)
library("factoextra") #Distance Matrix Computation and Visualization
library("corrplot") 

RVCdistdf <- read.csv("GeneticDistanceMatrix_RV-C.csv", row.names=1) #the pairwise genetic distance dataframe
RVCmetadf <-read.csv("RVC-PCA_metadata.csv", row.names=1) #the metadata associated to sequences/individuals
RVCmetadf[RVCmetadf ==''] <- NA
RVCmetadf$AgeGroup <- as.factor(RVCmetadf$AgeGroup)
RVCmetadf$Genotype <- as.factor(RVCmetadf$Genotype)
RVCmetadf$ZipCode <- as.factor(RVCmetadf$ZipCode)
RVCmetadf$Sex <- as.factor(RVCmetadf$Sex)
RVCmetadf$COLLECTION <- as.factor(substr(RVCmetadf$CollectionDate, 1, 7))
RVCmetadf$Clinica <- as.factor(RVCmetadf$Clinica)
RVCmetadf$Age <- as.numeric(RVCmetadf$Age)
RVCmetadf$RV_Ct_value <- as.numeric(RVCmetadf$RV_Ct_value)
```

Calculate the eigenvalues and principal component analysis: 
```{r PCA calculation, echo=TRUE}
#PCA analysis
res.RVC <- PCA(RVCdistdf, scale.unit = TRUE, ncp = 7)
eig.val <- get_eigenvalue(res.RVC)

#To evaluate the number of Principal Components we can see the Scree Plot, which is the plot of the eigenvalues ordered from largest to the smallest.
fviz_eig(res.RVC, addlabels = TRUE, ylim = c(0, 50))
```

Analysis of quality and contribution of individuals: 
```{r plot of individuals, echo=TRUE}
fviz_pca_ind(res.RVC, col.ind = "cos2", pointsize = "cos2", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), labelsize=2)

#To visualize the contribution of individuals to the first two principal components:
fviz_contrib(res.RVC, choice = "ind", axes = 1:2, labelsize=2) + theme(text = element_text(size=7))
```

Constructing a dendrogram and making the hierarchical clustering:
```{r Hierarchical Clustering, echo=TRUE}
library("dendextend") #cut tree into groups
library("svglite") #save plots
RVCdm<-dist(RVCdistdf, method = 'euclidean') 
RVChc<-hclust(RVCdm, method="complete") # simple dendrogram

#evaluating the number of clusters such that the total intra-cluster variation (total within-cluster sum of square, WSS) is minimized:
fviz_nbclust(RVCdistdf, FUNcluster=hcut, method="wss") + geom_vline(xintercept = 7, linetype = 2)

cut_cmp <- cutree(RVChc, k = 7) #applying the clustering by wss
plot(RVChc, hang=-1, cex=0.2) 
#plot with cluster rectangles within the clusters:
rect.hclust(RVChc, k=7, border=2:4)

#save plot of dendrogram:
#svglite(filename = "RVC_HC_PCA_euclidean.svg")
#plot(RVChc, hang=-1, cex=0.2) 
#rect.hclust(RVChc, k=7, border=2:4) #if I want to see the clusters identified by a rectangle
#dev.off()
```


Evaluate the association of metadata with the clusters by hierarchical clustering:
```{r anaysis of hierarchical clustering associated with metadata, echo=TRUE}
#make a dataframe with the clusters and evaluating association with metadata using G test of independence between groups:
library("dplyr")
library("tidyverse")
library("ggpubr") #boxplot with points
library("rstatix") #kruskal adn wilcox tests
library("rcompanion") #Pairwise tests of independence
library("AMR") #G-test
RVC_cluster <- merge(RVCmetadf, cut_cmp, by = 0)
RVC_cluster$y<-as.factor(RVC_cluster$y)

g.test(RVC_cluster$Clinica, RVC_cluster$y)
X1 = pairwiseNominalIndependence(as.matrix(table(RVC_cluster$Clinica, RVC_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X1, threshold  = 0.05)
ggplot(RVC_cluster, aes(x=Clinica, fill=y)) + geom_bar(position = "fill")

g.test(RVC_cluster$ZipCode, RVC_cluster$y)
X2 = pairwiseNominalIndependence(as.matrix(table(RVC_cluster$ZipCode, RVC_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X2, threshold  = 0.05)
ggplot(RVC_cluster, aes(x=ZipCode, fill=y)) + geom_bar(position = "fill") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

X3 = pairwiseNominalIndependence(as.matrix(table(RVC_cluster$Sex, RVC_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X3, threshold  = 0.05)
ggplot(RVC_cluster, aes(x=Sex, fill=y)) + geom_bar(position = "fill")

X4 = pairwiseNominalIndependence(as.matrix(table(RVC_cluster$AgeGroup, RVC_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3) 
cldList(p.adj.Gtest ~ Comparison, data= X4, threshold  = 0.05)
ggplot(RVC_cluster, aes(x=AgeGroup, fill=y)) + geom_bar(position = "fill") + scale_x_discrete(limits = c("less5", "5-17", "18-65", "more65"))

RVC_cluster$Year <- as.factor(substr(RVC_cluster$CollectionDate, 1, 4))
X5 = pairwiseNominalIndependence(as.matrix(table(RVC_cluster$Year, RVC_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X5, threshold  = 0.05)
ggplot(RVC_cluster, aes(x=Year, fill=y)) + geom_bar(position = "fill")

#Further analysis with continuous variables Age:
#Age
res.kruskal <- RVC_cluster %>% kruskal_test(Age ~ y) #Kruskal-Wallis Test
res.kruskal
#Wilcoxonâ€™s test to calculate pairwise comparisons between group:
pwc <- RVC_cluster %>% 
  wilcox_test(Age ~ y, p.adjust.method = "bonferroni") 
pwc

#visualization: box plots with p-values:
pwc <- pwc %>% add_xy_position(x = "y")
ggboxplot(RVC_cluster, x = "y", y = "Age") +
  stat_pvalue_manual(pwc, hide.ns = TRUE) +
  labs(
    subtitle = get_test_label(res.kruskal, detailed = TRUE),
    caption = get_pwc_label(pwc)
    )
```

