---
title: "RVA_PCA_analysis"
author: "Stephanie Goya"
date: "2023-12-18"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Upload the dataframes and alignment (in FASTA) to calculate the pairwise genetic distance matrix:
```{r dataframes, echo=TRUE}
library("ape")
library("seqinr") 
library("FactoMineR") 
library("factoextra") 
library("corrplot") 

RVAdistdf <- read.csv("RVAdistdf.csv", row.names=1) #the pairwise genetic distance dataframe
RVAmetadf <-read.csv("RVA-PCA_metadata.csv", row.names=1) #the metadata associated to sequences/individuals
RVAmetadf[RVAmetadf ==''] <- NA
RVAmetadf$AgeGroup <- as.factor(RVAmetadf$AgeGroup)
RVAmetadf$Genotype <- as.factor(RVAmetadf$Genotype)
RVAmetadf$ZIPCODE <- as.factor(RVAmetadf$ZIPCODE)
RVAmetadf$sex <- as.factor(RVAmetadf$sex)
RVAmetadf$COLLECTION <- as.factor(RVAmetadf$COLL_MONTH)
RVAmetadf$Conclusion <- as.factor(RVAmetadf$Conclusion)
RVAmetadf$age <- as.numeric(RVAmetadf$age)
RVAmetadf$Ct <- as.numeric(RVAmetadf$Ct)
```

Calculate the eigenvalues and principal component analysis: 
```{r PCA calculation, echo=TRUE}
#calculating the PCA
res.RVA <- PCA(RVAdistdf, scale.unit = TRUE, ncp = 4)
eig.val <- get_eigenvalue(res.RVA)

#To help determining the number of Principal Components we can see the Scree Plot, which is the plot of the eigenvalues ordered from largest to the smallest.
fviz_eig(res.RVA, addlabels = TRUE, ylim = c(0, 50))
```

Analysis of quality and contribution of individuals: 
```{r plot of individuals, echo=TRUE}
fviz_pca_ind(res.RVA, col.ind = "cos2", pointsize = "cos2", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), labelsize=2)

#To visualize the contribution of individuals to the first two principal components:
fviz_contrib(res.RVA, choice = "ind", axes = 1:2, labelsize=2) + theme(text = element_text(size=7))
```

Evaluating the PCA by clustering based on metadata information:
```{r individuals colored by group, echo=TRUE}
#Individuals by Age Group
fviz_pca_ind(res.RVA, geom.ind = "point", col.ind = RVAmetadf$AgeGroup, palette = "lancet", addEllipses = TRUE, legend.title = "Age Group")

#Individuals by Symptoms
fviz_pca_ind(res.RVA, geom.ind = "point", col.ind = RVAmetadf$Conclusion, palette = "lancet", addEllipses = TRUE, legend.title = "Symptoms")

#Individuals by month and year of collection
fviz_pca_ind(res.RVA, geom.ind = "point", col.ind = RVAmetadf$COLLECTION, palette = "lancet", addEllipses = TRUE, legend.title = "Collection Month")

#Individuals by Zip Code
fviz_pca_ind(res.RVA, geom.ind = "point", col.ind = RVAmetadf$ZIPCODE, addEllipses = TRUE, legend.title = "Zip Code")

#Individuals by Gender
fviz_pca_ind(res.RVA, geom.ind = "point", col.ind = RVAmetadf$sex, palette = "lancet", addEllipses = TRUE, legend.title = "Sex")

```

Constructing a dendrogram and making the hierarchical clustering:
```{r Hierarchical Clustering, echo=TRUE}
library("dendextend")
library("svglite")
RVAdm<-dist(RVAdistdf, method = 'euclidean') 
RVAhc<-hclust(RVAdm, method="complete") # simple dendrogram

#evaluating the number of clusters con average silhouette:
fviz_nbclust(RVAdistdf, FUNcluster=hcut, method="silhouette", k.max = 10)
fviz_nbclust(RVAdistdf, FUNcluster=hcut, method="wss")

cut_cmp <- cutree(RVAhc, k = 4) #applying the clustering by wss
plot(RVAhc, hang=-1, cex=0.2) 
#plot with cluster rectangles within the clusters:
rect.hclust(RVAhc, k=4, border=2:4)

#save plot of dendrogram:
#svg(filename = "RVA_HC_PCA_euclidean.svg")
#plot(RVAhc, hang=-1, cex=0.2) 
#rect.hclust(RVAhc, k=4, border=2:4) #if I want to see the clusters identified by a rectangle
#dev.off()
```


Evaluate the association of metadata with the clusters by hierarchical clustering:
```{r anaysis of hierarchical clustering associated with metadata, echo=TRUE}
#make a dataframe with the clusters and evaluating association with metadata using G test of independence between groups:
library("dplyr")
library("rcompanion")
library("tidyverse")
library("ggpubr")
library("rstatix")
library("AMR")
RVA_cluster <- merge(RVAmetadf, cut_cmp, by = 0)
RVA_cluster$y<-as.factor(RVA_cluster$y)

g.test(RVA_cluster$Conclusion, RVA_cluster$y)
X1 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster$Conclusion, RVA_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X1, threshold  = 0.05)
ggplot(RVA_cluster, aes(x=Conclusion, fill=y)) + geom_bar(position = "fill")

g.test(RVA_cluster$ZIPCODE, RVA_cluster$y)
X2 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster$ZIPCODE, RVA_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X2, threshold  = 0.05)
ggplot(RVA_cluster, aes(x=ZIPCODE, fill=y)) + geom_bar(position = "fill") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

X3 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster$sex, RVA_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X3, threshold  = 0.05)
ggplot(RVA_cluster, aes(x=sex, fill=y)) + geom_bar(position = "fill")

X4 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster$AgeGroup, RVA_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3) 
cldList(p.adj.Gtest ~ Comparison, data= X4, threshold  = 0.05)
ggplot(RVA_cluster, aes(x=AgeGroup, fill=y)) + geom_bar(position = "fill") + scale_x_discrete(limits = c("less5", "5-17", "18-65", "more65"))

#Further analysis with continuous variables age and Ct:
#AGE
res.kruskal <- RVA_cluster %>% kruskal_test(age ~ y) #Kruskal-Wallis Test
res.kruskal
#Wilcoxon’s test to calculate pairwise comparisons between group:
pwc <- RVA_cluster %>% 
  wilcox_test(age ~ y, p.adjust.method = "bonferroni") 
pwc

#visualization: box plots with p-values:
pwc <- pwc %>% add_xy_position(x = "y")
ggboxplot(RVA_cluster, x = "y", y = "age") +
  stat_pvalue_manual(pwc, hide.ns = TRUE) +
  labs(
    subtitle = get_test_label(res.kruskal, detailed = TRUE),
    caption = get_pwc_label(pwc)
    )

#Ct
res2.kruskal <- RVA_cluster %>% kruskal_test(Ct ~ y) #Kruskal-Wallis Test
res2.kruskal
#Wilcoxon’s test to calculate pairwise comparisons between group:
pwc2 <- RVA_cluster %>% 
  wilcox_test(Ct ~ y, p.adjust.method = "bonferroni") 
pwc2

#visualization: box plots with p-values:
pwc2 <- pwc2 %>% add_xy_position(x = "y")
ggboxplot(RVA_cluster, x = "y", y = "Ct") +
  stat_pvalue_manual(pwc2, hide.ns = TRUE) +
  labs(
    subtitle = get_test_label(res2.kruskal, detailed = TRUE),
    caption = get_pwc_label(pwc2)
    )

```

