---
title: "RVA_PCA_analysis"
author: "Stephanie Goya"
date: "2024-06-21"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Upload the dataframes and alignment (in FASTA) to calculate the pairwise genetic distance matrix:

```{r dataframes, echo=TRUE}
library("ape") #pairwise distance from DNA sequences
library("seqinr") #read alignment 
library("FactoMineR") #Principal Component Analysis (PCA)
library("factoextra") #Distance Matrix Computation and Visualization
library("corrplot")

RVAdistdf <- read.csv("GeneticDistanceMatrix_RV-A_20240621.csv", row.names=1) #the pairwise genetic distance dataframe
RVAmetadf <-read.csv("RVA-PCA_metadata_20240621.csv", row.names=1) #the metadata associated to sequences/individuals
RVAmetadf[RVAmetadf ==''] <- NA
RVAmetadf$AgeGroup <- as.factor(RVAmetadf$AgeGroup)
RVAmetadf$Genotype <- as.factor(RVAmetadf$Genotype)
RVAmetadf$ZipCode <- as.factor(RVAmetadf$ZipCode)
RVAmetadf$Sex <- as.factor(RVAmetadf$Sex)
RVAmetadf$COLLECTION <- as.factor(substr(RVAmetadf$CollectionDate, 1, 7))
RVAmetadf$Clinica <- as.factor(RVAmetadf$Symptoms)
RVAmetadf$Age <- as.numeric(RVAmetadf$Age)
RVAmetadf$RV_Ct_value <- as.numeric(RVAmetadf$RV_Ct_value)
```

Calculate the eigenvalues and principal component analysis:

```{r PCA calculation, echo=TRUE}
#PCA analysis
res.RVA <- PCA(RVAdistdf, scale.unit = TRUE, ncp = 5)
eig.val <- get_eigenvalue(res.RVA)

#To evaluate the number of Principal Components we can see the Scree Plot, which is the plot of the eigenvalues ordered from largest to the smallest.
fviz_eig(res.RVA, addlabels = TRUE, ylim = c(0, 50))
```

Analysis of quality and contribution of individuals:

```{r plot of individuals, echo=TRUE}
fviz_pca_ind(res.RVA, col.ind = "cos2", pointsize = "cos2", gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"), labelsize=2)

#To visualize the contribution of individuals to the first two principal components:
fviz_contrib(res.RVA, choice = "ind", axes = 1:2, labelsize=2) + theme(text = element_text(size=7))
```

Constructing a dendrogram and making the hierarchical clustering:

```{r Hierarchical Clustering, echo=TRUE}
library("dendextend") #cut tree into groups
library("svglite") #save plots
RVAdm<-dist(RVAdistdf, method = 'euclidean') 
RVAhc<-hclust(RVAdm, method="complete") # simple dendrogram

#evaluating the number of clusters such that the total intra-cluster variation (total within-cluster sum of square, WSS) is minimized:
fviz_nbclust(RVAdistdf, FUNcluster=hcut, method="wss") + geom_vline(xintercept = 4, linetype = 2)

cut_cmp <- cutree(RVAhc, k = 4) #applying the clustering by wss
plot(RVAhc, hang=-1, cex=0.2) 
#plot with cluster rectangles within the clusters:
rect.hclust(RVAhc, k=4, border=2:4)

#save plot of dendrogram:
#svglite(filename = "RVA_HC_PCA_euclidean_TEST.svg")
#plot(RVAhc, hang=-1, cex=0.2) 
#rect.hclust(RVAhc, k=4, border=2:4) #to see the clusters identified by a rectangle
#dev.off()
```

Evaluate the association of metadata with the clusters by hierarchical clustering:

```{r anaysis of hierarchical clustering associated with metadata, echo=TRUE}
#make a dataframe with the clusters and evaluating association with metadata using G test of independence between groups:
library("dplyr")
library("tidyverse")
library("ggpubr") #boxplot with points
library("rstatix") #kruskal adn wilcox tests
library("rcompanion") #Pairwise tests of independence
library("AMR") #G-test
RVA_cluster_pre <- merge(RVAmetadf, cut_cmp, by = 0)
RVA_cluster_pre$y<-as.factor(RVA_cluster_pre$y)
RVA_cluster <- RVA_cluster_pre[!is.na(RVA_cluster_pre$Age),]
RVA_cluster_clinica <- RVA_cluster[!is.na(RVA_cluster$Clinica),]
RVA_cluster_zipcode <- RVA_cluster[!is.na(RVA_cluster$ZipCode),]
RVA_cluster_sex <- RVA_cluster[!is.na(RVA_cluster$Sex),]

g.test(RVA_cluster_clinica$Clinica, RVA_cluster_clinica$y)
X1 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster_clinica$Clinica, RVA_cluster_clinica$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X1, threshold  = 0.05)
ggplot(RVA_cluster_clinica, aes(x=Clinica, fill=y)) + geom_bar(position = "fill")

g.test(RVA_cluster_zipcode$ZipCode, RVA_cluster_zipcode$y)
X2 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster_zipcode$ZipCode, RVA_cluster_zipcode$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X2, threshold  = 0.05)
ggplot(RVA_cluster_zipcode, aes(x=ZipCode, fill=y)) + geom_bar(position = "fill") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) 

X3 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster_sex$Sex, RVA_cluster_sex$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X3, threshold  = 0.05)
ggplot(RVA_cluster_sex, aes(x=Sex, fill=y)) + geom_bar(position = "fill")

X4 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster$AgeGroup, RVA_cluster$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3) 
cldList(p.adj.Gtest ~ Comparison, data= X4, threshold  = 0.05)
ggplot(RVA_cluster, aes(x=AgeGroup, fill=y)) + geom_bar(position = "fill") + scale_x_discrete(limits = c("<5y", "5-18y", "19-65y", ">65y"))

RVA_cluster_pre$Year <- as.factor(substr(RVA_cluster_pre$CollectionDate, 1, 4))
X5 = pairwiseNominalIndependence(as.matrix(table(RVA_cluster_pre$Year, RVA_cluster_pre$y)), compare = "column", fisher = FALSE, chisq = FALSE, method = "bonferroni", digits = 3)
cldList(p.adj.Gtest ~ Comparison, data= X5, threshold  = 0.05)
ggplot(RVA_cluster_pre, aes(x=Year, fill=y)) + geom_bar(position = "fill")


#Further analysis with continuous variables Age and Ct:
#Age
res.kruskal <- RVA_cluster %>% kruskal_test(Age ~ y) #Kruskal-Wallis Test
res.kruskal
#Wilcoxonâ€™s test to calculate pairwise comparisons between group:
pwc <- RVA_cluster %>% 
  wilcox_test(Age ~ y, p.adjust.method = "bonferroni") 
pwc

#visualization: box plots with p-values:
pwc <- pwc %>% add_xy_position(x = "y")
ggboxplot(RVA_cluster, x = "y", y = "Age") +
  stat_pvalue_manual(pwc, hide.ns = TRUE) +
  labs(
    subtitle = get_test_label(res.kruskal, detailed = TRUE),
    caption = get_pwc_label(pwc)
    )
```
